src/udev/udevd.c

```diff
@@ -382,6 +383,56 @@ static int worker_lock_block_device(sd_device *dev, int *ret_fd) {
         return 1;
 }
 
+static int worker_mark_block_device_read_only(sd_device *dev) {
+        _cleanup_close_ int fd = -1;
+        const char *val;
+        int state = 1, r;
+
+        assert(dev);
+
+        if (!arg_blockdev_read_only)
+                return 0;
+
+        /* Do this only once, when the block device is new. If the device is later retriggered let's not
+         * toggle the bit again, so that people can boot up with full read-only mode and then unset the bit
+         * for specific devices only. */
+        if (!device_for_action(dev, DEVICE_ACTION_ADD))
+                return 0;
+
+        r = sd_device_get_subsystem(dev, &val);
+        if (r < 0)
+                return log_device_debug_errno(dev, r, "Failed to get subsystem: %m");
+
+        if (!streq(val, "block"))
+                return 0;
+
+        r = sd_device_get_sysname(dev, &val);
+        if (r < 0)
+                return log_device_debug_errno(dev, r, "Failed to get sysname: %m");
+
+        /* Exclude synthetic devices for now, this is supposed to be a safety feature to avoid modification
+         * of physical devices, and what sits on top of those doesn't really matter if we don't allow the
+         * underlying block devices to recieve changes. */
+        if (STARTSWITH_SET(val, "dm-", "md", "drbd", "loop"))
+                return 0;
```

Comment:
I was testing this on a VM, and it also marks /dev/zram0 as read-only. Any kind of whitelist/blacklist approach is inevitably going to miss various kinds of kernel device classes. Is there a way to know which block devices have an actual underlying device? Looking at my system, the presence of `/sys/class/block/<foo>/device` seems to be a good indicator.

Reply:
> `/sys/class/block/<foo>/device`

Oh, this doesn't work very well either. It apparently is possible to have a rw partition on a read-only block device. And I don't see a good way to distinguish /dev/vda1 from /dev/zram1.

But please add nbd and zram to the whitelist.

Reply:
it's a blacklist, not a whitelist... i'll add it to that one though

Reply:
unfortunately we can't really determine this automatically. There's a similar whitelist in the locking code btw, kinda sucks, but I have no better idea for now. We can certainly revisit that though eventually, if we figure out a better way to figure these out.